### PR TITLE
Directly store the sparse matrix in the SCS format

### PR DESCRIPTION
**Pull request by blegat**
Closes https://github.com/jump-dev/SCS.jl/issues/186

### CODE REVIEW

File: src/MOI_wrapper.jl [hunk 1]
```diff
@@ -313,43 +339,80 @@ orderidx(idx, s) = idx
 function orderidx(idx, s::MOI.PositiveSemidefiniteConeTriangle)
     sympackedUtoLidx(idx, s.side_dimension)
 end
-function MOIU.load_constraint(optimizer::Optimizer, ci::MOI.ConstraintIndex, f::MOI.VectorAffineFunction, s::MOI.AbstractVectorSet)
+
+# The SCS format is b - Ax ∈ cone
+function _load_terms(colptr, rowval, nzval, terms, offset)
+    @show terms
```
Comment (mlubin) [inline src/MOI_wrapper.jl:345]: Debugging code?

**Comment by kalmarek** [comment]
I can't really comment on the MOI part, but changes in `c_wrapper.jl` look ok. Also there are some lost `@show`s in `types.jl` ;)

File: src/MOI_wrapper.jl [hunk 2]
```diff
@@ -262,7 +291,7 @@ sympackedUtoL(x, n=sympackeddim(length(x))) = _sympackedto(x, n, trimap, (i, j)
 
 function sympackedUtoLidx(x::AbstractVector{<:Integer}, n)
     y = similar(x)
-    map = sympackedLtoU(1:sympackedlen(n), n)
+    map =
```
Comment (odow) [inline src/MOI_wrapper.jl:294]: What is this

File: src/MOI_wrapper.jl [hunk 3]
```diff
@@ -239,8 +261,15 @@ end
...
 end
-function MOIU.allocate_constraint(optimizer::Optimizer, f::F, s::S) where {F <: MOI.AbstractFunction, S <: MOI.AbstractSet}
-    return CI{F, S}(_allocate_constraint(optimizer.cone, f, s))
+function allocate_terms(colptr, terms)
+    for term in terms
+        colptr[term.scalar_term.variable_index.value + 1] += 1
+    end
+end
+function MOIU.allocate_constraint(optimizer::Optimizer, f::MOI.VectorAffineFunction{Float64}, s::S) where S <: MOI.AbstractSet
+    func = MOIU.canonical(f)
```
Comment (odow) [inline src/MOI_wrapper.jl:270]: This is a case for `is_canonical`

**Comment by odow** [comment]
> What's the improvement on the benchmarks?

Mainly a reduction in memory usage. 2x speedup in SCS scalar solve. Not sure why the generate slows by 2x though. All I did was swap SCS, and the `generate_scs_problem` doesn't use any SCS calls??? I tried multiple times and times are fairly stable.

### Before

```
  ─────────────────────────────────────────────────────────────────────────
                                  Time                   Allocations      
                          ──────────────────────   ───────────────────────
     Tot / % measured:         8.96s / 93.3%           3.72GiB / 100%     

 Section          ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────────
 SCS MOI scalar        1    5.95s  71.2%   5.95s   2.73GiB  73.5%  2.73GiB
   solve               1    4.32s  51.7%   4.32s   1.69GiB  45.4%  1.69GiB
   generate            1    1.63s  19.5%   1.63s   1.04GiB  28.1%  1.04GiB
 SCS MOI vector        1    1.55s  18.5%   1.55s    722MiB  19.0%   722MiB
   solve               1    1.40s  16.8%   1.40s    556MiB  14.6%   556MiB
   generate            1    146ms  1.75%   146ms    166MiB  4.36%   166MiB
 SCS direct            1    860ms  10.3%   860ms    287MiB  7.53%   287MiB
   solve               1    570ms  6.82%   570ms   38.5MiB  1.01%  38.5MiB
   generate            1    290ms  3.46%   290ms    248MiB  6.52%   248MiB
 ─────────────────────────────────────────────────────────────────────────
```

### After

```
 ─────────────────────────────────────────────────────────────────────────
                                  Time                   Allocations      
                          ──────────────────────   ───────────────────────
     Tot / % measured:         8.16s / 93.1%           2.12GiB / 100%     

 Section          ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────────
 SCS MOI scalar        1    5.00s  65.9%   5.00s   1.52GiB  71.6%  1.52GiB
   generate            1    2.98s  39.3%   2.98s   1.08GiB  51.1%  1.08GiB
   solve               1    2.02s  26.6%   2.02s    444MiB  20.5%   444MiB
 SCS MOI vector        1    1.43s  18.8%   1.43s    330MiB  15.2%   330MiB
   solve               1    1.29s  17.0%   1.29s    164MiB  7.56%   164MiB
   generate            1    133ms  1.75%   133ms    166MiB  7.64%   166MiB
 SCS direct            1    1.17s  15.4%   1.17s    287MiB  13.2%   287MiB
   solve               1    608ms  8.01%   608ms   38.5MiB  1.77%  38.5MiB
   generate            1    558ms  7.34%   558ms    248MiB  11.4%   248MiB
 ─────────────────────────────────────────────────────────────────────────
```

**Comment by odow** [comment]
@blegat is this ready to merge and tag?

**Comment by blegat** [comment]
Yes, let's merge so that it's used in SolverTests

**Pull request merged**